sootup.java.core/src/main/java/sootup/java/core/exceptions/ExceptionInferStmtVisitor.java

```diff
@@ -0,0 +1,159 @@
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseInvokeStmt(@Nonnull JInvokeStmt stmt) {
+    if (stmt.getInvokeExpr().isPresent()) {
+      Expr expr = stmt.getInvokeExpr().get();
+      expr.accept(exprVisitor);
+      result = result.addExceptions(exprVisitor.getResult(), hierarchy);
+    }
+  }
+
+  @Override
+  public void caseAssignStmt(@Nonnull JAssignStmt stmt) {
+    Value leftOp = stmt.getLeftOp();
+    Value rightOp = stmt.getRightOp();
+    // store in array
+    if (leftOp instanceof Ref) {
+      if (leftOp instanceof JArrayRef
+          && (leftOp.getType() instanceof UnknownType || leftOp.getType() instanceof ClassType)) {
+        result =
+            result.addException(
+                ExceptionInferResult.ExceptionType.ARRAY_STORE_EXCEPTION, hierarchy);
+      }
+      ((Ref) leftOp).accept(refVisitor);
+      result = result.addExceptions(refVisitor.getResult(), hierarchy);
+    }
+    if (rightOp instanceof Ref) {
+      ((Ref) rightOp).accept(refVisitor);
+      result = result.addExceptions(refVisitor.getResult(), hierarchy);
+    } else if (rightOp instanceof Expr) {
+      ((Expr) rightOp).accept(exprVisitor);
+      result = result.addExceptions(exprVisitor.getResult(), hierarchy);
+    }
+  }
+
+  @Override
+  public void caseIdentityStmt(@Nonnull JIdentityStmt stmt) {
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseEnterMonitorStmt(@Nonnull JEnterMonitorStmt stmt) {
+    result =
+        result.addException(ExceptionInferResult.ExceptionType.NUll_POINTER_EXCEPTION, hierarchy);
+  }
+
+  @Override
+  public void caseExitMonitorStmt(@Nonnull JExitMonitorStmt stmt) {
+    result =
+        result.addException(ExceptionInferResult.ExceptionType.NUll_POINTER_EXCEPTION, hierarchy);
+    result =
+        result.addException(
+            ExceptionInferResult.ExceptionType.ILLEGAL_MONITOR_STATE_EXCEPTION, hierarchy);
+  }
+
+  @Override
+  public void caseGotoStmt(@Nonnull JGotoStmt stmt) {
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseIfStmt(@Nonnull JIfStmt stmt) {
+    defaultCaseStmt(stmt);
+    // ConditionExpr has no implicit exceptions
+    /*Expr conditionExpr = stmt.getCondition();
+    conditionExpr.accept(exprVisitor);
+    result = result.addExceptions(exprVisitor.getResult(), hierarchy);*/
+  }
+
+  @Override
+  public void caseNopStmt(@Nonnull JNopStmt stmt) {
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseRetStmt(@Nonnull JRetStmt stmt) {
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseReturnStmt(@Nonnull JReturnStmt stmt) {
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseReturnVoidStmt(@Nonnull JReturnVoidStmt stmt) {
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseSwitchStmt(@Nonnull JSwitchStmt stmt) {
+    defaultCaseStmt(stmt);
+  }
+
+  @Override
+  public void caseThrowStmt(@Nonnull JThrowStmt stmt) {
+    defaultCaseStmt(stmt);
```

Comment:
You can define the ThrowStmt-specific logic here.

Reply:
We can get the inferred Exceptions from JThrowStmt per mightThrowExplicitly, because JThrowStmt tell us explicitly which exception it throws. Now I doesn't grasp all logics of Traptightener. Can we first maintain the structure align with old soot? If the differentiation is not used, I can merge them in branch TrapTightner.